### PR TITLE
Add note when item accessed from module via `m.i` rather than `m::i`. 

### PR DESCRIPTION
**Pull request by pnkfelix**
Add note when item accessed from module via `m.i` rather than `m::i`.

(I tried to make this somewhat future-proofed, in that the `UnresolvedNameContext` could be expanded in the future with other cases besides paths that are known to be modules.)

This supersedes PR #30356 ; since I'm responsible for a bunch of new code here, someone else should review it. :)

### CODE REVIEW

**Comment by pnkfelix** [comment]
gaar I posted this before running `make check` locally ... so of course it has tidy issues...

**Comment by GuillaumeGomez** [comment]
I'm still working on adding the two missing additional information.

**Comment by pnkfelix** [comment]
> I'm still working on adding the two missing additional information.

(and I think we do not need to block this PR on that change.)

**Comment by pnkfelix** [comment]
@GuillaumeGomez Just to be clear: The reason that I turned #30356  into this PR was that I was concerned that it was giving "bad hints" in some contexts.

 * I was concerned that #30356 was reporting an error mentioning a function call form like `m.f(...)` in source code contexts where there was no occurrence of `.f(...)`. 

I wrote up the [examples you noted](https://github.com/rust-lang/rust/pull/30356#issuecomment-165137689) as ways to illustrate cases where #30356 would give bad hints.

But I did *not* intend to imply something like "we should wait until we have solved *all* of these cases before we land anything."

My intention in this PR was to continue to solve the same narrow problem that #30356 was trying to solve, but to do it in a more targeted fashion that ensures that we only provide in hint about `m.i` or `m.f(..)` in cases where there actually *is* an `m.i` or `m.f(..)` present.

----

I definitely encourage you to try to come up with good error messages for other cases that this PR does not cover.  You should do that work and put it up as a separate PR.

File: src/librustc_resolve/lib.rs [hunk 1]
```diff
@@ -202,6 +202,12 @@ pub enum ResolutionError<'a> {
     AttemptToUseNonConstantValueInConstant,
 }
 
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub enum UnresolvedNameContext {
```
Comment (Manishearth) [inline src/librustc_resolve/lib.rs]: Could you add a doc comment on this so it's not confusing for future hackers?

(Also explain what the nodeid is)

**Comment by Manishearth** [comment]
r=me with nit

**Comment by frewsxcv** [comment]
This should also solve #22692 ?

**Comment by nikomatsakis** [comment]
@pnkfelix 

> The reason that I turned #30356 into this PR was that I was concerned that it was giving "bad hints" in some contexts.

Can you add tests that check that we do NOT provide bad hints? Hmm, I guess that's hard to do, isn't it?

**Comment by nikomatsakis** [comment]
@frewsxcv 

> This should also solve #22692 ?

I suspect not. This seems targeted at modules, not items found in impls on types (but resolve works in mysterious ways and constructs some pseudo-modules, so maybe it works sometimes).

**Comment by pnkfelix** [comment]
> Can you add tests that check that we do NOT provide bad hints? Hmm, I guess that's hard to do, isn't it?

I don't think I'm going to try to do that.  :)

**Comment by pnkfelix** [comment]
What the heck, a bunch of errors in `C:/bot/slave/auto-win-gnu-32-opt/build/src/llvm/lib/Target/Mips/MipsInstrInfo.h` ...?

**Pull request merged**